### PR TITLE
bpf: xdp: don't support GENEVE passthrough with DSR-Hybrid

### PR DESCRIPTION
**Pull request by julianwiedmann**
tail_nodeport_nat_egress_ipv4() doesn't have any support for XDP passthrough yet. See encap_geneve_dsr_opt4() for what would be needed to rewrite the outer headers. We also haven't considered how the reply traffic for non-DSR connections would get handled in the LB.

Therefore limit the encap-detection to full-DSR.

### CODE REVIEW

File: bpf/bpf_xdp.c [hunk 1]
```diff
@@ -110,7 +110,7 @@ int tail_lb_ipv4(struct __ctx_buff *ctx)
 			goto out;
 		}
 
-#if defined(ENABLE_DSR) && DSR_ENCAP_MODE == DSR_ENCAP_GENEVE
+#if defined(ENABLE_DSR) && !defined(ENABLE_DSR_HYBRID) && DSR_ENCAP_MODE == DSR_ENCAP_GENEVE
```
Comment (gentoo-root) [inline bpf/bpf_xdp.c:113]: Could you explain it further to me?

Do I understand correctly that the idea is to always hit [this branch](https://github.com/cilium/cilium/blob/1649229b1207e1c7f1db508e1856d7b9016d6fdf/bpf/lib/nodeport.h#L2950) and get the right treatment of the outer header at encap_geneve_dsr_opt4?

What would happen in the hybrid mode before this change? Are UDP packets also encapsulated into GENEVE? This is why we don't want to extract the inner header here?

What would happen after this change? All packets will be passed with ip4 pointing to the GENEVE UDP header, right? Doesn't nodeport_lb4 need to parse the actual inner packet?

Comment (julianwiedmann) [reply bpf/bpf_xdp.c:113]: > Could you explain it further to me?
> 
> Do I understand correctly that the idea is to always hit [this branch](https://github.com/cilium/cilium/blob/1649229b1207e1c7f1db508e1856d7b9016d6fdf/bpf/lib/nodeport.h#L2950) and get the right treatment of the outer header at encap_geneve_dsr_opt4?
> 

yep, the DSR path is the only one we initially cared about. I hadn't considered that DSR-Hybrid would be sending some traffic down the NAT path.

> What would happen in the hybrid mode before this change? Are UDP packets also encapsulated into GENEVE? This is why we don't want to extract the inner header here?
> 

For a setup with `ENABLE_DSR_HYBRID`, XDP would currently DNAT the inner UDP packet (if it's a service request) and then tail-call to `tail_nodeport_nat_egress_ipv4()`. Expecting it to update the outer GENEVE encap as needed (for instance change the outer `daddr`, based on an ipcache lookup for the inner `daddr`). But the code for such updates doesn't exist :).

We could make this decision at runtime instead - extract the inner L4 protocol, and only continue with "XDP passthrough" processing if the packet is DSR-eligible (`nodeport_uses_dsr()`). But right now I'm not aware that anyone would require such a mode.

> What would happen after this change? All packets will be passed with ip4 pointing to the GENEVE UDP header, right? Doesn't nodeport_lb4 need to parse the actual inner packet?

We essentially get the same behaviour as on v1.13. XDP only looks at the outer headers (GENEVE), tries to LB it (no match), tries to revSNAT it (no match), and eventually passes it up. `bpf_host`'s `from-netdev` tries the same thing. Eventually the geneve kernel driver strips off the encapsulation, and `from-overlay` load-balances the inner packet.

Comment (gentoo-root) [reply bpf/bpf_xdp.c:113]: Thanks for the detailed explanation!

**Review by gentoo-root — APPROVED** [review]

**Pull request merged**